packages/plugins/experimental/plugin-calls/src/state/call-swarm-synchronizer.ts

```diff
@@ -0,0 +1,264 @@
+import { isNonNullable } from '@dxos/util';
+
+import { codec, type TranscriptionState, type EncodedTrackName, type UserState } from '../types';
+
+export type CallState = {
+  /**
+   * All users states in the swarm.
+   */
+  users?: UserState[];
+  /**
+   * Self state inside swarm.
+   */
+  self?: UserState;
+
+  /**
+   * Swarm synchronized CRDT transcription. Last write wins.
+   */
+  transcription?: TranscriptionState;
+
+  raisedHand?: boolean;
+  speaking?: boolean;
+  joined?: boolean;
+
+  /**
+   * Tracks pushed by self to calls service.
+   */
+  tracks?: {
+    screenshare?: EncodedTrackName;
+    video?: EncodedTrackName;
+    audio?: EncodedTrackName;
+  };
+};
+
+export type CallSwarmSynchronizerParams = { networkService: NetworkService };
+
+/**
+ * Sends and receives state to/from Swarm network.
+ */
+export class CallSwarmSynchronizer extends Resource {
+  private readonly _state: CallState = { transcription: { enabled: false, lamportTimestamp: { id: '', version: 0 } } };
+  public readonly stateUpdated = new Event<CallState>();
+
+  private readonly _networkService: NetworkService;
+
+  private _identityKey?: string = undefined;
+  private _deviceKey?: string = undefined;
+  private _displayName?: string = undefined;
+
+  private _stream?: Stream<SwarmResponse> = undefined;
+  private _roomId?: PublicKey = undefined;
+  private _sendStateTask?: DeferredTask = undefined;
+
+  constructor({ networkService }: CallSwarmSynchronizerParams) {
+    super();
+    this._networkService = networkService;
+  }
+
+  /**
+   * @internal
+   */
+  _getState() {
+    return this._state;
+  }
+
+  setRaisedHand(raisedHand: boolean) {
+    if (this._state.raisedHand === raisedHand) {
+      return;
+    }
+
+    this._state.raisedHand = raisedHand;
+    this._notifyAndSchedule();
+  }
+
+  setSpeaking(speaking: boolean) {
+    if (this._state.speaking === speaking) {
+      return;
+    }
+
+    this._state.speaking = speaking;
+    this._notifyAndSchedule();
+  }
+
+  setJoined(joined: boolean) {
+    if (this._state.joined === joined) {
+      return;
+    }
+
+    this._state.joined = joined;
+    this._notifyAndSchedule();
+  }
+
+  setTracks(tracks: { screenshare?: EncodedTrackName; video?: EncodedTrackName; audio?: EncodedTrackName }) {
+    if (
+      tracks.video === this._state.tracks?.video &&
+      tracks.audio === this._state.tracks?.audio &&
+      tracks.screenshare === this._state.tracks?.screenshare
+    ) {
+      return;
+    }
+
+    this._state.tracks = { ...this._state.tracks, ...tracks };
+    this._notifyAndSchedule();
+  }
+
+  setTranscription(transcription: TranscriptionState) {
+    const currentTranscription = this._state.transcription;
+    this._state.transcription = {
+      ...currentTranscription,
+      ...transcription,
+      lamportTimestamp: LamportTimestampCrdt.increment(currentTranscription?.lamportTimestamp ?? {}) as any,
```

Comment:
why as any?

Reply:
Typescript for some reason is not working properly here